pkg/endpoint/endpoint.go

```diff
@@ -1630,6 +1637,100 @@ func (e *Endpoint) APICanModifyConfig(n models.ConfigurationMap) error {
+//     operation will be done in the background and 'regenTriggered'
+//     will always be 'false'.
+//
+//   - bwm - the bandwidth manager used to update the bandwidth policy for this
+//     endpoint.
+//
+//   - resolveMetadata - the metadata resolver that will be used to retrieve this
+//     endpoint's metadata.
+func (e *Endpoint) metadataResolver(ctx context.Context,
+	restoredEndpoint, blocking bool,
+	baseLabels labels.Labels,
+	bwm bandwidth.Manager,
+	resolveMetadata MetadataResolverCB) (regenTriggered bool, err error) {
+	if !e.K8sNamespaceAndPodNameIsSet() {
+		e.Logger(resolveLabels).Debug("Namespace and Pod are not set")
+		return false, nil
+	}
+
+	// copy the base labels into this local variable
+	// so that we don't override 'baseLabels'.
+	controllerBaseLabels := labels.NewFrom(baseLabels)
+
+	ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
+
+	pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+	if err != nil {
+		e.Logger(resolveLabels).WithError(err).Warning("Unable to fetch kubernetes labels")
+		// If we were unable to fetch the k8s endpoints then
+		// we will mark the endpoint with the init identity.
+		if !restoredEndpoint {
+			// Only mark the endpoint with the 'init' identity if we are not
+			// restoring the endpoint from a restart.
+			identityLabels := labels.Labels{
+				labels.IDNameInit: labels.NewLabel(labels.IDNameInit, "", labels.LabelSourceReserved),
+			}
+			regenTriggered := e.UpdateLabels(ctx, labels.LabelSourceAny, identityLabels, info, true)
+			if blocking {
+				return regenTriggered, err
+			}
+		}
+		return false, err
+	}
+
+	// Merge the labels retrieved from the 'resolveMetadata into the base
+	// labels.
+	controllerBaseLabels.MergeLabels(identityLabels)
+
+	e.SetPod(pod)
+	e.SetK8sMetadata(cp)
+	e.UpdateNoTrackRules(func(_, _ string) (noTrackPort string, err error) {
+		po, _, _, _, _, err := resolveMetadata(ns, podName)
+		if err != nil {
+			return "", err
+		}
+		value, _ := annotation.Get(po, annotation.NoTrack, annotation.NoTrackAlias)
+		return value, nil
+	})
+	e.UpdateVisibilityPolicy(func(_, _ string) (proxyVisibility string, err error) {
+		po, _, _, _, _, err := resolveMetadata(ns, podName)
+		if err != nil {
+			return "", err
+		}
+		value, _ := annotation.Get(po, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
+		return value, nil
+	})
+	e.UpdateBandwidthPolicy(bwm, func(ns, podName string) (bandwidthEgress string, err error) {
+		_, _, _, _, annotations, err := resolveMetadata(ns, podName)
+		if err != nil {
+			return "", err
+		}
+		return annotations[bandwidth.EgressBandwidth], nil
+	})
+
+	// Since 'controllerBaseLabels' were fetched from k8s then we
+	// should only replace the labels that have their source as
+	// 'k8s' otherwise we will risk on replacing other labels that
+	// were added from other sources.
+	source := labels.LabelSourceK8s
+	if len(baseLabels) != 0 {
+		source = labels.LabelSourceAny
+	}
```

Comment:
I feel like the `len(baseLabels) != 0` logic contradicts the comment. If there's base labels, then k8s labels can also overwrite the base labels. Then we reintroduce one of the bugs that this patch aims to solve?

Should this be doing two calls or passing down two sets of labels where the base labels use `LabelSourceAny` (and potentially happen after / override the k8s labels?)?

Reply:
The comment is indeed confusing, is this better?
> 	// If 'baseLabels' are not set then 'controllerBaseLabels' only contains
>	// labels from k8s. Thus, we should only replace the labels that have their
>	// source as 'k8s' otherwise we will risk on replacing other labels that
>	// were added from other sources.

Reply:
That context helps, but I think we still have an issue?

If `baseLabels` *is* set, then any labels we learn from Kubernetes could overwrite labels from other sources.

Reply:
Ah, I think that implicitly if `baseLabels` is set then the endpoint is being created, so updating the identity cannot overwrite labels, because there are no labels to overwrite(?)

In that case I think it's logically correct, just confusing.

Maybe a particularly weird case would be, what if the API labels during endpoint create include a label with source k8s? Will Cilium sanitize that input, could it conflict with k8s labels?

Reply:
> Ah, I think that implicitly if `baseLabels` is set then the endpoint is being created, so updating the identity cannot overwrite labels, because there are no labels to overwrite(?)

Correct.
> 
> In that case I think it's logically correct, just confusing.
> 
> Maybe a particularly weird case would be, what if the API labels during endpoint create include a label with source k8s? Will Cilium sanitize that input, could it conflict with k8s labels?

If the endpoint create contains let's say "k8s:foo=bar" and then the metadataResolver, which has the k8s source, fetches the data with "foo=baz" then the final labels will be "k8s:foo=baz". Just because they have the same source.

If we had "cni:foo=bar", then the final labels would be "cni:foo=bar". The k8s won't replace the initial labels.

Reply:
I see. That sounds like the right behaviour.

If the underlying `UpdateLabels` implementation provides that guarantee then I still don't understand the need for this if conditional, because labels from k8s should only ever have the k8s source, so they should never overwrite labels from other sources. Is this just defensive coding in case something changes in future?

I looked around but I couldn't find a specific test that validated the difference in behaviour for this conditional based on the input labels.

Reply:
Answer: The `UpdateLabels` API in the endpoint is quirky in a few ways that makes this specific pattern necessary:
- If you use `UpdateLabels(..., SourceAny, newLabels, ...)`, it is equivalent to deleting all labels in the endpoint and adding all labels in the `newLabels` set.
- If you use `UpdateLabels(..., SourceK8s, newLabels, ...)`, it is equivalent to deleting all labels in the endpoint with k8s source and adding the labels in the `newLabels` set.

For endpoint creation, we know that some labels may come from the API, and we know that the endpoint has no labels currently, so we can replace all labels. After endpoint creation (including during restore), resolving the metadata from k8s will only present new k8s labels, and will *not* retrieve the source of truth for non-k8s labels. Therefore, if we use `SourceAny` with the metadata resolver, it could end up deleting all non-k8s labels.

I think that the `UpdateLabels` API here is overly complex and is what causes some of these complications, but this code is actually correct and necessary as-is.

As follow-ups I think it would be nice to rethink how the Endpoint's label-setting APIs work to make them less confusing.